### PR TITLE
PrintTable: recurse over hashes

### PR DESCRIPTION
**Pull request by lacostej**
This implements 3 things:
- allow to recurse over hashes
- allow to mask some some entries (instead of hiding them)
- do not simulate asking for the entry when in test of ci mode if one is in ask: false mode.

If for example mask_keys is rejected, the rest should be recovered :)

### CODE REVIEW

File: lib/fastlane_core/print_table.rb [hunk 1]
```diff
@@ -5,26 +5,47 @@ class << self
...
-          next if hide_keys.include?(config_item.key)
 
-          rows << [config_item.key, value]
-        end
+        options = config.nil? ? {} : config.values
+        rows = self.collect_rows(options: options, hide_keys: hide_keys.map(&:to_s), prefix: '')
 
         params = {}
-        params[:rows] = rows
+        params[:rows] = limit_row_size rows
```
Comment (KrauseFx) [inline lib/fastlane_core/print_table.rb]: Please surround with `(rows)`

File: lib/fastlane_core/print_table.rb [hunk 2]
```diff
@@ -5,26 +5,47 @@ class << self
...
         params[:title] = title if title
+        # params[:style] = {width: 80} # broken with 1.4.5
+        # workaround in the meantime
 
         puts ""
         puts Terminal::Table.new(params)
         puts ""
 
-        return params
+        params
```
Comment (KrauseFx) [inline lib/fastlane_core/print_table.rb]: Please leave the `return`

File: lib/fastlane_core/string_filters.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  # The last characters will be replaced with the <tt>:omission</tt> string (defaults to "...")
+  # for a total length not exceeding <tt>length</tt>:
+  #
+  #   'And they found that many people were sleeping better.'.truncate(25, omission: '... (continued)')
+  #   # => "And they f... (continued)"
+  def truncate(truncate_at, options = {})
+    return dup unless length > truncate_at
+
+    omission = options[:omission] || '...'
+    length_with_room_for_omission = truncate_at - omission.length
```
Comment (KrauseFx) [inline lib/fastlane_core/string_filters.rb]: :disappointed: so much code

Comment (lacostej) [inline lib/fastlane_core/string_filters.rb]: if we fix https://github.com/lacostej/fastlane_core/blob/2b36de7a3c159cc2437a511d874f9ccbd1a3f793/lib/fastlane_core/print_table.rb#L15, then we can probably get rid of that extra-code altogether.

Comment (lacostej) [inline lib/fastlane_core/string_filters.rb]: That's also why I've let it in an extra module

Comment (KrauseFx) [inline lib/fastlane_core/string_filters.rb]: What is broken?

Comment (KrauseFx) [inline lib/fastlane_core/string_filters.rb]: What doesn't work/

File: lib/fastlane_core/print_table.rb (hunk 2 above)
Comment (lacostej) [inline lib/fastlane_core/print_table.rb]: Out of curiosity, why ? Other methods don't typically use return. Is there a specific style guideline to follow ?

File: lib/fastlane_core/string_filters.rb (hunk 3 above)
Comment (lacostej) [inline lib/fastlane_core/string_filters.rb]: I wanted to use params[:style] = {width: 80}, which I thought would handle line-wrapping. After testing, it doesn't and won't do it. See https://github.com/tj/terminal-table/issues/1#issuecomment-4100874

So I removed the comments

**Comment by asfalcone** [comment]
@lacostej This PR is a bit old but looks to be in pretty good shape. If you're still interested in us merging this PR, please let me know if there is anything else you want to add or just squash the commits. 🚀

**Comment by mfurtak** [comment]
This got merged (directly to master :fearful: - sorry!) as part of 5369e93de0c029d3e62f91b0382ac8a21e3ddb0f

**Pull request closed without merging**